flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/StateChangeLogger.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.state.changelog;
+
+import org.apache.flink.core.memory.DataOutputViewStreamWrapper;
+import org.apache.flink.util.ExceptionUtils;
+import org.apache.flink.util.function.BiConsumerWithException;
+import org.apache.flink.util.function.ThrowingConsumer;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.Map;
+
+interface StateChangeLogger<State, Namespace> {
+    static <Namespace, State, StateElement> Iterator<StateElement> loggingIterator(
```

Comment:
Let me answer in the reverse order:
3. Yes, those methods are needed because of `Iterator.remove` (and `Map.Entry.setValue`). These actions should be logged.
2. `MapState` **does** use these methods, but it also uses `StateChangeLogger` to log other state modifications
1. So to me, these methods are closely related to the interface functionality

However, I'm also fine with extracting it into something like `StateChangeLoggingUtils` if that sounds good to you.

Reply:
I've moved  `loggingMapEntry` to `ChangelogMapState` as discussed offline, as well as `loggingIterable` (as it's only used in map state currently and can easily be extracted);
I've also replaced `loggingIterator` with separate `StateChangeLoggingIterator` class.